apps/emqx_bridge_snowflake/src/emqx_bridge_snowflake_channel_client.erl

```diff
@@ -0,0 +1,329 @@
+        setup_pool_id => SetupPoolId,
+        channel_name => ChannelName,
+        path => OpenChannelPath
+    }),
+    Resp = ?MODULE:do_open_channel(SetupPoolId, Req, RequestTTL, MaxRetries),
+    maybe
+        {ok, 200, _, BodyRaw} ?= emqx_bridge_http_connector:transform_result(Resp),
+        #{<<"next_continuation_token">> := ContinuationToken} =
+            Response0 = emqx_utils_json:decode(BodyRaw),
+        ?tp(debug, "snowflake_streaming_open_channel_success", #{
+            channel_name => ChannelName,
+            response => Response0
+        }),
+        State = State0#{
+            ?continuation_token := ContinuationToken,
+            ?open_channel_error := ?undefined
+        },
+        {ok, State}
+    else
+        {ok, Code, RespHeaders, RespBodyRaw} ->
+            RespBody =
+                case emqx_utils_json:safe_decode(RespBodyRaw) of
+                    {ok, JSONErr} -> JSONErr;
+                    {error, _} -> RespBodyRaw
+                end,
+            Error =
+                {error, #{
+                    reason => <<"unexpected_response_opening_channel">>,
+                    response => {Code, RespHeaders, RespBody}
+                }},
+            State1 = State0#{?open_channel_error := Error},
+            {Error, State1};
+        Response ->
+            Error =
+                {error, #{
+                    reason => <<"unexpected_response_opening_channel">>,
+                    response => Response
+                }},
+            State1 = State0#{?open_channel_error := Error},
+            {Error, State1}
+    end.
+
+%% Internal export exposed ONLY for mocking
+do_open_channel(HTTPPool, Req, RequestTTL, MaxRetries) ->
+    ehttpc:request(HTTPPool, put, Req, RequestTTL, MaxRetries).
+
+do_append_rows(State0, Records) ->
+    #{
+        ?id := Id,
+        ?append_rows_path := AppendRowsPath0,
+        ?continuation_token := ContinuationToken,
+        ?channel_name := ChannelName,
+        ?jwt_config := JWTConfig,
+        ?request_ttl := RequestTTL,
+        ?max_retries := MaxRetries,
+        ?write_pool_id := WritePoolId
+    } = State0,
+    JWTToken = emqx_connector_jwt:ensure_jwt(JWTConfig),
+    maybe
+        Headers = [
+            {<<"Content-Type">>, <<"application/json">>},
+            {<<"Accept">>, <<"application/json">>},
+            {<<"User-Agent">>, <<"emqx">>},
+            {<<"Authorization">>, [<<"Bearer ">>, JWTToken]}
+        ],
+        Body = lists:map(fun(R) -> [emqx_utils_json:encode(R), $\n] end, Records),
+        AppendRowsPath = iolist_to_binary([
+            AppendRowsPath0,
+            <<"?continuationToken=">>,
+            uri_quote(ContinuationToken)
+        ]),
+        Req = {AppendRowsPath, Headers, Body},
+        ?tp(debug, "snowflake_streaming_append_rows_request", #{
+            write_pool_id => WritePoolId,
+            channel => ChannelName,
+            path => AppendRowsPath,
+            records => Records
+        }),
```

Comment:
Nit. Those expressions are basically infallible, and having them inside `maybe` adds extra cognitive load.